akka-stream/src/main/scala/akka/stream/impl/FoldResourceSinkAsync.scala

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright (C) 2015-2019 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.stream.impl
+
+import akka.Done
+import akka.annotation.InternalApi
+import akka.dispatch.ExecutionContexts.sameThreadExecutionContext
+import akka.stream.ActorAttributes.SupervisionStrategy
+import akka.stream._
+import akka.stream.impl.Stages.DefaultAttributes
+import akka.stream.stage._
+
+import scala.concurrent.{ Future, Promise }
+import scala.util.{ Failure, Success, Try }
+import scala.util.control.NonFatal
+
+/**
+ * INTERNAL API
+ */
+@InternalApi private[akka] final class FoldResourceSinkAsync[T, S](
+  create: () ⇒ Future[S],
+  write:  (S, T) ⇒ Future[Unit],
+  close:  (S) ⇒ Future[Unit]) extends GraphStageWithMaterializedValue[SinkShape[T], Future[Done]] {
+  val in = Inlet[T]("FoldResourceSinkAsync.in")
+  override val shape = SinkShape(in)
+  override def initialAttributes: Attributes = DefaultAttributes.foldResourceSinkAsync
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, Future[Done]) = {
+    val promise = Promise[Done]()
+
+    val createLogic = new GraphStageLogic(shape) with InHandler {
+      lazy val decider = inheritedAttributes.mandatoryAttribute[SupervisionStrategy].decider
+
+      private implicit def ec = materializer.executionContext
+
+      private var blockingStream: Option[S] = None
+
+      private var writeFail: Option[Throwable] = None
+
+      private val createdCallback = getAsyncCallback[Try[S]] {
+        case Success(resource) ⇒
+          blockingStream = Some(resource)
+          pull(in)
+
+        case Failure(t) ⇒ failStage(t)
+      }.invokeWithFeedback _
+
+      private val errorHandler: PartialFunction[Throwable, Unit] = {
+        case NonFatal(ex) ⇒ decider(ex) match {
+          case Supervision.Stop ⇒
+            writeFail = Some(ex)
+            failStage(ex)
+          case Supervision.Restart ⇒ restartResource()
+          case Supervision.Resume  ⇒ pull(in)
+        }
+      }
+
+      private val pushCallback = getAsyncCallback[Try[Unit]] {
+        case Success(_) ⇒ pull(in)
+        case Failure(t) ⇒ errorHandler(t)
+      }.invoke _
+
+      override def preStart(): Unit = createResource()
+
+      override def onPush(): Unit =
+        blockingStream match {
+          case Some(resource) ⇒
+            try {
+              write(resource, grab(in)).onComplete(pushCallback)(sameThreadExecutionContext)
```

Comment:
There is an optimization to do for this (per element) future return value, where we can check if the future is already completed, and not go via the async callback in that case, that should make it efficient enough that we do not need the other graph stage impl at all, but can use this for both and just return a `Future.successful(resource)` for the calls of the non async version.

I don't think it is worth doing the optimization for opening/closing though.

Reply:
Something like this?

```
      override def onPush(): Unit =
        blockingStream match {
          case Some(resource) ⇒
            try {
              val f = write(resource, grab(in))
              f.value match {
                case Some(t) ⇒ t match {
                  case Failure(ex) ⇒ errorHandler(ex)
                  case _           ⇒ pull(in)
                }

                case _ ⇒ f.onComplete(pushCallback)(sameThreadExecutionContext)
              }
            } catch errorHandler
          case None ⇒
          // we got a pull but there is no open resource, we are either
          // currently creating/restarting then the pull will be triggered when creating the
          // resource completes, or shutting down and then the push does not matter anyway
        }
```

Reply:
Yes, could be worth extracting that into a `writeCompleted` method to make the logic stays the same for both paths in the future.